### DISCUSSION TITLE
Importing Bot_Token in to the flow.json

### DISCUSSION THREAD

**Discussion by ROF13ThFloor** [windkh/node-red-contrib-telegrambot#368] [Q&A] [answered]
Hi , 

I'm trying to import the flow with node-red-contrib-telegrambot to the cloud with Node-Red npm package .

It seems that when I'm trying to export a flow with the telegram bot node, token are not exported to the flow.json file.

I've tried to import token by ENV but it's not working on the cloud with node-red npm package 

does any one has idea about this problem?

**Comment by windkh** [comment, marked as answer]
_(Drafted by Claude, Anthropic's AI coding assistant, after a code-level review of the bot-node config-resolution path. The recommendation below is Claude's, posted by the maintainer.)_

@ROF13ThFloor — apologies for the very late reply. There are actually **two separate problems** in the same shape here that need different fixes. Both are now addressed.

### 1. The token is not in `flows.json` — by design (Node-RED framework behaviour, not the plugin)

Node-RED treats every bot's `token` field as a *credential* property and intentionally segregates it into **`flows_cred.json`**, encrypted with your local credential secret. That is why you don't see the token in `flows.json` after deploy. This is a Node-RED design choice that the plugin can't override — and shouldn't, since credentials sharing the same file as the flow logic would defeat the encryption.

For cloud / automated / git-managed deployments, two patterns work:

**Pattern A — commit the encrypted credentials file with a deterministic credential secret.** Set the credential secret in `settings.js`:

```js
module.exports = {
    credentialSecret: process.env.NODE_RED_CREDENTIAL_SECRET,
    // ...
}
```

Then `flows_cred.json` becomes reproducible across deploys as long as the env var is the same. Commit both files; the token survives.

**Pattern B — populate the credentials file at boot from your secrets store.** Render `flows_cred.json` at container start from whatever your platform uses for secrets (Doppler, AWS Secrets Manager, GCP Secret Manager, Kubernetes Secrets, etc.), then let Node-RED read it normally.

The detailed walkthrough — full `settings.js`, Dockerfile patterns, troubleshooting — is in **[DEPLOYMENT.md](https://github.com/windkh/node-red-contrib-telegrambot/blob/master/DEPLOYMENT.md)** (added in V17.4.0).

### 2. `{env.get("X")}` in the token field didn't actually resolve — plugin bug, fixed in V17.3.1

What you tried with the env-var route had a *second* problem hiding behind the credentials one. The plugin's `getUserNames` / `getChatIds` (and the token resolver) used to read `{env.get("X")}` correctly but then only handle the result as a single value — so a perfectly reasonable `CHATIDS="123,456"` env var resolved to the **string** `"123,456"` instead of `[123, 456]`. That broke the chat-ID authorisation list silently.

[V17.3.1](https://github.com/windkh/node-red-contrib-telegrambot/releases/tag/V17.3.1) splits string results on commas (and `parseInt`s for chat IDs), matching the literal-comma branch. Arrays returned from `flow.get` / `global.get` still pass through unchanged. Same fix applies if you put a comma-separated value in any of the env-var fields.

### Putting it together for the cloud / npm deployment case

Concrete recipe for the case you described:

1. Upgrade to V17.4.5 (latest as of today) — gets you both fixes plus the recent stability work on issue #442.
2. In your container / cloud setup, set `NODE_RED_CREDENTIAL_SECRET` to a stable, secret value (do *not* check this one into git).
3. In `settings.js`: `credentialSecret: process.env.NODE_RED_CREDENTIAL_SECRET`.
4. Either:
    - Commit `flows_cred.json` alongside `flows.json` (it's encrypted with the secret), OR
    - Set `TELEGRAM_BOT_TOKEN=...` as an env var and put `{env.get("TELEGRAM_BOT_TOKEN")}` in the token field. The plugin resolves env-get tokens at startup.

If you can share what platform you're deploying to (Heroku, Railway, plain Docker, k8s...) I can point at the exact env-var configuration step for that platform. Closing this as effectively answered — please open it back up or comment if either fix doesn't behave as documented.
